hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetAsyncDiskService.java

```diff
@@ -359,6 +371,89 @@ public void run() {
+                + ": GenerationStamp not matched, existing replica is "
+                + Block.toString(infoByBlockId) + " in removeReplicaFromMem.");
+          }
+          return false;
+        }
+
+        FsVolumeImpl v = (FsVolumeImpl)info.getVolume();
+        if (v == null) {
+          LOG.error("Failed to delete replica " + block.getLocalBlock()
+              +  ". No volume for this replica " + info + " in removeReplicaFromMem.");
+          return false;
+        }
+
+        try {
+          File blockFile = new File(info.getBlockURI());
+          if (blockFile != null && blockFile.getParentFile() == null) {
+            LOG.error("Failed to delete replica " + block.getLocalBlock()
+                +  ". Parent not found for block file: " + blockFile
+                + " in removeReplicaFromMem.");
+            return false;
+          }
+        } catch(IllegalArgumentException e) {
+          LOG.warn("Parent directory check failed; replica {} is " +
+              "not backed by a local file in removeReplicaFromMem.", info);
+        }
+
+        if (!this.volume.getStorageID().equals(v.getStorageID())) {
+          LOG.error("Failed to delete replica " + block.getLocalBlock()
+              +  ". Appear different volumes, oldVolume=" + this.volume + " and newVolume=" + v
+              +  " for this replica in removeReplicaFromMem.");
+          return false;
+        }
+
+        ReplicaInfo removing = fsdatasetImpl.volumeMap.remove(block.getBlockPoolId(),
+            block.getLocalBlock());
+        fsdatasetImpl.addDeletingBlock(block.getBlockPoolId(), removing.getBlockId());
+        LOG.debug("Block file {} is to be deleted", removing.getBlockURI());
+        datanode.getMetrics().incrBlocksRemoved(1);
+        if (removing instanceof ReplicaInPipeline) {
+          ((ReplicaInPipeline) removing).releaseAllBytesReserved();
+        }
+      }
+
+      if (volume.isTransientStorage()) {
+        RamDiskReplicaTracker.RamDiskReplica replicaInfo =
+            fsdatasetImpl.ramDiskReplicaTracker.getReplica(block.getBlockPoolId(),
+                block.getLocalBlock().getBlockId());
+        if (replicaInfo != null) {
+          if (!replicaInfo.getIsPersisted()) {
+            datanode.getMetrics().incrRamDiskBlocksDeletedBeforeLazyPersisted();
+          }
+          fsdatasetImpl.ramDiskReplicaTracker.discardReplica(replicaInfo.getBlockPoolId(),
+              replicaInfo.getBlockId(), true);
+        }
+      }
+
+      // If a DFSClient has the replica in its cache of short-circuit file
+      // descriptors (and the client is using ShortCircuitShm), invalidate it.
+      datanode.getShortCircuitRegistry().processBlockInvalidation(
+          new ExtendedBlockId(block.getLocalBlock().getBlockId(), block.getBlockPoolId()));
+
+      // If the block is cached, start uncaching it.
+      fsdatasetImpl.cacheManager.uncacheBlock(
+          block.getBlockPoolId(), block.getLocalBlock().getBlockId());
```

Comment:
Maybe we can use local variable to store blockPoolId and LocalBlock to clean the code. such as
```
final String blockPoolId = block.getBlockPoolId();
final Block localBlock = block.getLocalBlock();

fsdatasetImpl.cacheManager.uncacheBlock(blockPoolId, localBlock.getBlockId());
```